src/Interpreters/convertFieldToType.cpp

```diff
@@ -356,7 +356,7 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
             for (size_t i = 0; i < src_arr_size; ++i)
             {
                 res[i] = convertFieldToType(src_arr[i], element_type);
-                if (res[i].isNull() && !element_type.isNullable())
+                if (res[i].isNull() && !canContainNull(element_type))
```

Comment:
Github acted up or I messed up, but my comment is lost.

So this condition and the one below basically serves the same purpose: so set `have_unconvertible_element` only when we have issues in conversion. However the completely different phrasing (`A AND !B` and `!A OR B`) makes them a bit confusing. I know it is not strictly connected to the PR, so feel free to merge it without changing it, but if you are up for the change and willing to do it, it would be nice.

Reply:
Let's use `A AND !B` in all places